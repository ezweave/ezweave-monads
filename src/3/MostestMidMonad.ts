interface IMonad<A> {
  bind<B>(callback: (value?: A) => Monad<B>): IMonad<B>;
  map<B>(callback: (value?: A) => B): IMonad<B>;
}

abstract class Monad<A> implements IMonad<A> {
  abstract bind<B>(fn: (value?: A) => IMonad<B>): IMonad<B>;
  abstract map<B>(callback: (value?: A) => B): IMonad<B>;

  public static some<A>(value: A) {
    return new MostestMidMonad<A>(value);
  }

  public static none() {
    return new MostestMidMonad<undefined>();
  }
}

export class MostestMidMonad<A> extends Monad<A> {
  myValue?: A;

  constructor(t?: A) {
    super();
    if (t) {
      this.myValue = t;
    }
  }

  public bind<B>(callback: (value?: A) => IMonad<B>) {
    return callback(this.myValue);
  }

  public map<B>(callback: (value?: A) => B) {
    return MostestMidMonad.some(callback(this.myValue));
  }
}
