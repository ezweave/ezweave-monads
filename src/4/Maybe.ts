interface IMonad<A> {
  bind<B>(callback: (value?: A) => Monad<B>): IMonad<B>;
  map<B>(callback: (value?: A) => B): IMonad<B>;
}

abstract class Monad<A> implements IMonad<A> {
  abstract bind<B>(fn: (value?: A) => IMonad<B>): IMonad<B>;
  abstract map<B>(callback: (value?: A) => B): IMonad<B>;

  public static some<A>(value: A) {
    return new Maybe<A>(value);
  }

  public static none() {
    return new Maybe<undefined>();
  }
}

export class Maybe<A> extends Monad<A> {
  myValue?: A;

  constructor(t?: A) {
    super();
    if (t) {
      this.myValue = t;
    }
  }

  public bind<B>(callback: (value?: A) => IMonad<B>) {
    return callback(this.myValue);
  }

  public map<B>(callback: (value?: A) => B) {
    return this.isSome() ? Maybe.some(callback(this.myValue)) : Maybe.none();
  }

  public isSome(): boolean {
    return !this.isNone();
  }

  public isNone(): boolean {
    return this.myValue === null || this.myValue === undefined;
  }
}
