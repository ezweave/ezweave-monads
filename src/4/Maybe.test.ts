import { toLower } from 'lodash';

import { Maybe } from './Maybe';
import { MostestMidMonad } from '@ezweave/3/MostestMidMonad';

describe(Maybe, () => {
  it('does nothing if there is no value', () => {
    const monad = Maybe.none().map(toLower);

    expect(monad).toMatchSnapshot();
  });
  it('does not break if a function returns nothing', () => {
    const monad = Maybe.some(1)
      .map((n) => n! % 2 === 0 ? n : undefined)
      .map((n) => n.toFixed());

    expect(monad).toMatchSnapshot();
  });
  it('would break if we used a different Monad', () => {
    expect(() => MostestMidMonad.some(1)
      .map((n) => n! % 2 === 0 ? n : undefined)
      // @ts-ignore
      .map((n) => n.toFixed())).toThrowErrorMatchingSnapshot();
  })
});
