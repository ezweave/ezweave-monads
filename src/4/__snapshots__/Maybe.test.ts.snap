// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maybe does not break if a function returns nothing 1`] = `Maybe {}`;

exports[`Maybe does nothing if there is no value 1`] = `Maybe {}`;

exports[`Maybe would break if we used a different Monad 1`] = `"Cannot read properties of undefined (reading 'toFixed')"`;
